File: frontend/src/api/events.ts
import { apiClient } from './client'

export type ContactActionPayload = {
  actionType: string
  contentName?: string | null
  sourcePage?: string
  metaEventId?: string
  fbp?: string | null
  fbc?: string | null
  userAgent?: string
  eventSourceUrl?: string
}

export const sendContactAction = (payload: ContactActionPayload) =>
  apiClient.post('/events/contact-action', {
    action_type: payload.actionType,
    content_name: payload.contentName,
    source_page: payload.sourcePage,
    meta_event_id: payload.metaEventId,
    fbp: payload.fbp,
    fbc: payload.fbc,
    user_agent: payload.userAgent,
    event_source_url: payload.eventSourceUrl,
  })

File: backend/app/api/events.py
from fastapi import APIRouter, Request, status

from app.schemas.event import ContactActionCreate
from app.services.meta_capi import send_contact_action_event


router = APIRouter(prefix="/events", tags=["events"])


def _client_ip(request: Request) -> str | None:
    forwarded_for = request.headers.get("x-forwarded-for")
    if forwarded_for:
        return forwarded_for.split(",")[0].strip()

    if request.client:
        return request.client.host

    return None


@router.post("/contact-action", status_code=status.HTTP_202_ACCEPTED)
def contact_action(payload: ContactActionCreate, request: Request):
    payload.client_ip_address = payload.client_ip_address or _client_ip(request)
    payload.user_agent = payload.user_agent or request.headers.get("user-agent")
    send_contact_action_event(payload)
    return {"status": "ok"}
